src/common/mmap.c

```diff
@@ -492,47 +501,100 @@ util_range_unregister(const void *addr, size_t len)
 }
 
 /*
- * util_range_is_pmem -- return true if entire range
- * is persistent memory
+ * util_range_is_pmem -- return true if entire range is persistent memory
  */
 int
-util_range_is_pmem(const void *addrp, size_t len)
+util_range_is_pmem(const void *addr, size_t len)
 {
-	LOG(10, "addr %p len %zu", addrp, len);
+	LOG(10, "addr %p len %zu", addr, len);
 
-	uintptr_t addr = (uintptr_t)addrp;
-	int retval = 1;
+	if (len > UINTPTR_MAX - (uintptr_t)addr) {
+		len = UINTPTR_MAX - (uintptr_t)addr;
+		LOG(4, "limit len to %zu to not get beyond address space", len);
+	}
+
+	uintptr_t begin = (uintptr_t)addr;
+	uintptr_t end = (uintptr_t)addr + len;
+
+	LOG(4, "begin 0x%016" PRIxPTR " end 0x%016" PRIxPTR, begin, end);
+
+	int retval = -1;
 
 	util_rwlock_rdlock(&Mmap_list_lock);
 
-	do {
-		struct map_tracker *mt = util_range_find(addr, len);
-		if (mt == NULL) {
-			LOG(4, "address not found 0x%016" PRIxPTR, addr);
+	struct map_tracker *mt;
+	SORTEDQ_FOREACH(mt, &Mmap_list, entry) {
+		if (mt->end_addr <= begin) {
+			LOG(4, "skipping all mapped ranges before given range");
+			continue;
+		}
+		if (mt->base_addr >= end) {
+			LOG(4, "ignoring all mapped ranges beyond given range");
+			break;
+		}
+
+		if (begin < mt->base_addr) {
+			LOG(4, "untracked address");
 			retval = 0;
 			break;
 		}
 
-		LOG(10, "range found - begin 0x%016" PRIxPTR
-				" end 0x%016" PRIxPTR,
+#ifdef _WIN32
+		/*
+		 * This is only needed for Windows, where all the mappings
+		 * are tracked. On Linux, only DAX mappings are on the list.
+		 */
+		if (mt->type != PMEM_MAP_DAX) {
+			LOG(4, "tracked range [%p, %p) is not direct mapped",
 				mt->base_addr, mt->end_addr);
+			retval = 0;
+			break;
+		}
 
-		if (mt->base_addr > addr) {
-			LOG(10, "base address doesn't match: "
-				"0x%" PRIxPTR " > 0x%" PRIxPTR,
-					mt->base_addr, addr);
+		/*
+		 * If there is a gap between the given region that we process
+		 * currently and the mapped region in our tracking list, we
+		 * need to process the gap by taking the long route of asking
+		 * MM for each page in that range.
+		 */
+		if (begin < mt->base_addr &&
+		    !util_is_direct_mapped(begin, mt->base_addr)) {
+			LOG(4, "untracked range [%p, %p) is not direct mapped",
+				begin, mt->base_addr);
 			retval = 0;
 			break;
 		}
+#endif
 
-		uintptr_t map_len = mt->end_addr - addr;
-		if (map_len > len)
-			map_len = len;
-		len -= map_len;
-		addr += map_len;
-	} while (len > 0);
+		retval = 1;
+
+		/* push our begin to reflect what we have already processed */
+		begin = mt->end_addr;
+	}
+
+#ifdef _WIN32
+	/*
+	 * If we still have a range to verify, check with MM if the entire
+	 * region is direct mapped.
+	 */
+	if (begin < end && !util_is_direct_mapped(begin, end)) {
```

Comment:
Maybe we should just implement `util_is_direct_mapped` on linux (as always true if it we were to retain the current logic).